### ISSUE TITLE
Compiler does not detect impossible situation with default function parameters

### ISSUE THREAD

**Issue by victornoel** [Microsoft/TypeScript#15797]
**TypeScript Version:**  2.3.2

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
function f({ a = false, b = '/' } = {}) {
}

f(true);
f('lol');
f(5);
```

**Expected behavior:**

The 3 calls to `f` should fail at compile time.

**Actual behavior:**

The calls do not fail.

**Comment:**

I suppose that typescript consider that the argument to the function is an object. And thus the parameter available inside the function body will simply be this same object but with two extra properties `a` and `b` set to their default values.
It makes sense, but it seems completely counter-intuitive to the programmer expectations :)

**Comment by mhegazy** [comment]
The signature of this function is `f(arg: { a?: boolean, b?: string }): void;`, since this type only has optional members, it is technically assignable from any type (as long as it does not have an `a`  or a `b` with conflicting properties). This is another manifestation of https://github.com/Microsoft/TypeScript/issues/7485.

**Comment by victornoel** [comment]
Closing this in favour of #7485. Thanks :)
